paasta_tools/contrib/check_orphans.py

```diff
@@ -0,0 +1,211 @@
+    blacklisted_services = set(yaml.load(service_blacklist_path))
+    # Dump ZK
+    zk_data = get_zk_data(blacklisted_services)
+    zk_instance_data = read_from_zk_data(zk_data)
+
+    hosts = {x[0] for x in zk_instance_data}
+
+    # Dump Nerve configs from each host via xinetd
+    results = asyncio.get_event_loop().run_until_complete(gather_files(hosts))
+
+    nerve_instance_data, not_found_hosts = read_nerve_files(results)
+
+    # Filter out anything that we couldn't get a nerve config for
+    zk_instance_data_filtered = {x for x in zk_instance_data if x[0] not in not_found_hosts}
+
+    logger.info('zk_instance_data (unfiltered) len: {}'.format(len(zk_instance_data)))
+    logger.info('zk_instance_data (filtered) len: {}'.format(len(zk_instance_data_filtered)))
+    logger.info('nerve_instance_data len: {}'.format(len(nerve_instance_data)))
+
+    return zk_instance_data_filtered, nerve_instance_data
+
+
+def check_orphans(
+    zk_instance_data: Set[InstanceTuple],
+    nerve_instance_data: Set[InstanceTuple],
+) -> ExitCode:
+    orphans = zk_instance_data - nerve_instance_data
+
+    # groupby host
+    orphans_by_host: DefaultDict[str, List[Tuple[int, str]]] = defaultdict(list)
+    for orphan in orphans:
+        orphans_by_host[orphan.host].append((orphan.port, orphan.service))
+
+    # collisions
+    instance_by_addr: DefaultDict[Tuple[str, int], Set[str]] = defaultdict(set)
+    for nerve_inst in nerve_instance_data:
+        instance_by_addr[(nerve_inst.host, nerve_inst.port)].add(nerve_inst.service)
+
+    collisions: List[str] = []
+    for zk_inst in zk_instance_data:
+        nerve_services = instance_by_addr[(zk_inst.host, zk_inst.port)]
+        if len(nerve_services) >= 1 and zk_inst.service not in nerve_services:
+            collisions.append(
+                f'[{zk_inst.host}:{zk_inst.port}] {zk_inst.service} collides with {nerve_services}',
+            )
+
+    if collisions:
+        logger.warning('Collisions found! Traffic is being misrouted!')
+        print('\n'.join(collisions))
+        return ExitCode.COLLISIONS
+    elif orphans:
+        logger.warning('{} orphans found'.format(len(orphans)))
+        print(dict(orphans_by_host))
+        return ExitCode.ORPHANS
+    else:
+        logger.info(
+            'No orphans found out of {} service registrations seen'.format(len(zk_instance_data)),
+        )
+        return ExitCode.OK
+
+
+def main() -> ExitCode:
+    logging.basicConfig(level=logging.INFO)
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--service-blacklist-path',
+        default=DEFAULT_SERVICE_BLACKLIST_PATH,
+        help="Path to a file that contains services to ignore",
+    )
+    args = parser.parse_args()
+
+    zk_instance_data, nerve_instance_data = get_instance_data(args.service_blacklist_path)
+
+    return check_orphans(zk_instance_data, nerve_instance_data)
+
+
+if __name__ == '__main__':
+    sys.exit(main().value)
```

Comment:
what is up with the `.value` here? It isn't already just a number?

Reply:
nope. it's an `Enum`. https://docs.python.org/3/library/enum.html#programmatic-access-to-enumeration-members-and-their-attributes